sdk/monitor/azure-monitor-query/src/main/java/com/azure/monitor/query/LogsAsyncClient.java

```diff
@@ -0,0 +1,228 @@
+        return queryLogsWithResponse(new LogsQueryOptions(workspaceId, query, timeSpan))
+                .map(Response::getValue);
+    }
+
+    /**
+     * Returns all the Azure Monitor logs matching the given query in the specified workspaceId.
+     * @param options The query options.
+     * @return The logs matching the query.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<LogsQueryResult>> queryLogsWithResponse(LogsQueryOptions options) {
+        return withContext(context -> queryLogsWithResponse(options, context));
+    }
+
+    /**
+     * Returns all the Azure Monitor logs matching the given batch of queries in the specified workspaceId.
+     * @param workspaceId The workspaceId where the batch of queries should be executed.
+     * @param queries A batch of Kusto queries.
+     * @param timeSpan The time period for which the logs should be looked up.
+     * @return A collection of query results corresponding to the input batch of queries.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<LogsQueryBatchResultCollection> queryLogsBatch(String workspaceId, List<String> queries,
+                                                               QueryTimeSpan timeSpan) {
+        LogsQueryBatch logsQueryBatch = new LogsQueryBatch();
+        queries.forEach(query -> logsQueryBatch.addQuery(workspaceId, query, timeSpan));
+        return queryLogsBatchWithResponse(logsQueryBatch).map(Response::getValue);
+    }
+
+    /**
+     * Returns all the Azure Monitor logs matching the given batch of queries.
+     * @param logsQueryBatch {@link LogsQueryBatch} containing a batch of queries.
+     * @return A collection of query results corresponding to the input batch of queries.@return
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<LogsQueryBatchResultCollection>> queryLogsBatchWithResponse(LogsQueryBatch logsQueryBatch) {
+        return queryLogsBatchWithResponse(logsQueryBatch, Context.NONE);
+    }
+
+    Mono<Response<LogsQueryBatchResultCollection>> queryLogsBatchWithResponse(LogsQueryBatch logsQueryBatch, Context context) {
+        BatchRequest batchRequest = new BatchRequest();
+        AtomicInteger id = new AtomicInteger();
+
+        List<LogQueryRequest> requests = logsQueryBatch.getQueries()
+                .stream()
+                .map(query -> {
+                    QueryBody body = new QueryBody(query.getQuery())
+                            .setWorkspaces(query.getWorkspaceNames())
+                            .setAzureResourceIds(query.getAzureResourceIds())
+                            .setQualifiedNames(query.getQualifiedWorkspaceNames())
+                            .setWorkspaceIds(query.getWorkspaceIds());
+
+                    return new LogQueryRequest()
+                            .setId(String.valueOf(id.incrementAndGet()))
```

Comment:
Why atomic reference here? This is a serial stream I believe java mandates use of atomic reference in lambda but never understood why?

Reply:
When using lambdas, all references should be final or effectively final. Using an int and incrementing it would result in a mutable field which is not allowed. So, atomic integer is used to keep the reference final but increment the value that it holds.